common/lib/util/eventemitter.ts

```diff
@@ -2,81 +2,83 @@ import * as Utils from './utils';
+				listeners.splice(idx, 1);
+			}
+			/* If events object has an event name key with no listeners then
+					remove the key to stop the list growing indefinitely */
+			if (eventFilter && (listeners.length === 0)) {
+				delete targetListeners[targetListenersIndex][eventFilter];
+			}
+		} else if (Utils.isObject(listeners)) {
+			/* events */
+			for (eventName in listeners) {
+				if (Object.prototype.hasOwnProperty.call(listeners, eventName) && Utils.isArray(listeners[eventName])) {
+					removeListener([listeners], listener, eventName);
+				}
+			}
+		}
+	}
+}
 
-var EventEmitter = (function() {
+class EventEmitter {
+	any: Array<Function>;
+	events: Record<string, Array<Function>>;
+	anyOnce: Array<Function>;
+	eventsOnce: Record<string, Array<Function>>;
 
-	/* public constructor */
-	function EventEmitter() {
+	constructor() {
 		this.any = [];
 		this.events = Object.create(null);
 		this.anyOnce = [];
 		this.eventsOnce = Object.create(null);
 	}
 
-	/* Call the listener, catch any exceptions and log, but continue operation*/
-	function callListener(eventThis, listener, args) {
-		try {
-			listener.apply(eventThis, args);
-		} catch(e) {
-			Logger.logAction(Logger.LOG_ERROR, 'EventEmitter.emit()', 'Unexpected listener exception: ' + e + '; stack = ' + (e && e.stack));
-		}
-	}
-
-	/**
-	 * Remove listeners that match listener
-	 * @param targetListeners is an array of listener arrays or event objects with arrays of listeners
-	 * @param listener the listener callback to remove
-	 * @param eventFilter (optional) event name instructing the function to only remove listeners for the specified event
-	 */
-	function removeListener(targetListeners, listener, eventFilter) {
-		var listeners, idx, eventName, targetListenersIndex;
-
-		for (targetListenersIndex = 0; targetListenersIndex < targetListeners.length; targetListenersIndex++) {
-			listeners = targetListeners[targetListenersIndex];
-			if (eventFilter) { listeners = listeners[eventFilter]; }
-
-			if (Utils.isArray(listeners)) {
-				while ((idx = Utils.arrIndexOf(listeners, listener)) !== -1) {
-					listeners.splice(idx, 1);
-				}
-				/* If events object has an event name key with no listeners then
-				   remove the key to stop the list growing indefinitely */
-				if (eventFilter && (listeners.length === 0)) {
-					delete targetListeners[targetListenersIndex][eventFilter];
-				}
-			} else if (Utils.isObject(listeners)) {
-				/* events */
-				for (eventName in listeners) {
-					if (hasOwnProperty.call(listeners, eventName) && Utils.isArray(listeners[eventName])) {
-						removeListener([listeners], listener, eventName);
-					}
-				}
-			}
-		}
-	}
-
 	/**
 	 * Add an event listener
 	 * @param event (optional) the name of the event to listen to
 	 *        if not supplied, all events trigger a call to the listener
 	 * @param listener the listener to be called
 	 */
-	EventEmitter.prototype.on = function(event, listener) {
+	on (event: string, listener: Function) {
```

Comment:
This is restricting to a single signature, (even, listener), but on() is overloaded, eg you can call with a listener but no event. With this that call will fail typechecking.

With typescript the idiomatic way of doing this is multiple function signatures, where the last one is the 'active' one which is the union of all the others, and the others are basically just documentation -- eg:

```
on(listener: Function): void;
on(event: null | string | string[], listener: Function): void;
on(event: null | string | string[] | Function, listener?: Function): void {
```

Reply:
32f8d04